src/main/java/org/opensearch/knn/jni/JNIService.java

```diff
@@ -88,7 +88,55 @@ public static long loadIndex(String indexPath, Map<String, Object> parameters, S
             return FaissService.loadIndex(indexPath);
         }
 
-        throw new IllegalArgumentException("LoadIndex not supported for provided engine");
+        throw new IllegalArgumentException("LoadIndex not supported for provided engine : " + engineName);
+    }
+
+    /**
+     * Determine if index contains shared state. Currently, we cannot do this in the plugin because we do not store the
+     * model definition anywhere. Only faiss supports indices that have shared state. So for all other engines it will
+     * return false.
+     *
+     * @param indexAddr address of index to be checked.
+     * @param engineName name of engine
+     * @return true if index requires shared index state; false otherwise
+     */
+    public static boolean isSharedIndexStateRequired(long indexAddr, String engineName) {
+        if (KNNEngine.FAISS.getName().equals(engineName)) {
+            return FaissService.isSharedIndexStateRequired(indexAddr);
+        }
+
+        return false;
+    }
+
+    /**
+     * Initialize the shared state for an index
+     *
+     * @param indexAddr address of the index to initialize from
+     * @param engineName name of engine
+     * @return Address of shared index state address
+     */
+    public static long initSharedIndexState(long indexAddr, String engineName) {
+        if (KNNEngine.FAISS.getName().equals(engineName)) {
+            return FaissService.initSharedIndexState(indexAddr);
+        }
+
+        throw new IllegalArgumentException("InitSharedIndexState not supported for provided engine : " + engineName);
+    }
+
+    /**
+     * Set the index state for an index
+     *
+     * @param indexAddr address of index to set state for
+     * @param shareIndexStateAddr address of shared state to be set
+     * @param engineName name of engine
+     */
+    public static void setSharedIndexState(long indexAddr, long shareIndexStateAddr, String engineName) {
+        if (KNNEngine.FAISS.getName().equals(engineName)) {
```

Comment:
this is third similar check for the engine name. Should we make it part of the engine class with default value of false, and only for faiss is will be true? Maybe in one of next PR as part of refactoring

Reply:
Yes, we could think about this. Probably in another refactoring PR.